### ISSUE TITLE
Add time integration constraint for quaternions

### ISSUE THREAD

**Issue by hongkai-dai** [RobotLocomotion/drake#14956]
In trajectory optimization, we often use Euler integration like
```
q[n+1] = q[n] + qdot[n+1] * dt
```
This is valid for robot prismatic/revolute joint angles, but not correct for a quaternion. If `z[n+1]` is a valid unit quaternions, then we know that its time-derivative `zdot[n+1]` has to be perpendicular to `z[n]` (due to the unit-length constraint on quaternions). Hence `z[n+1] - zdot[n+1]*dt` is on the tangent surface of the unit sphere, and this tangent surface only intersects the unit sphere at `z[n+1]`. Since `z[n]` is also a unit quaternion, the only solution of `z[n] = z[n+1] - zdot[n+1]*dt` is that `z[n] = z[n+1]`.

We need a different integration scheme for quaternions. I propose the following idea:
1. We first get the average angular velocity in the time interval [t[n], t[n+1]], we denote this average angular velocity as ω̅. This average angular velocity computed from zero-order hold (ω̅ = ω[n]), or from linear interpolation ω̅ = (ω[n] + ω[n+1]) / 2
2. We then impose the constraint z[n+1] = z(ω̅) ⊗ z[n], where z(ω̅) = [cos(|ω̅|*dt/2), sin(|ω̅|*dt/2) * ω̅/|ω̅|] representing the quaternion of rotating about the fixed axis ω̅/|ω̅| with an angle |ω̅|*dt, ⊗ is the product between two quaternions.

This approach guarantees to satisfy the unit length constraint for quaternions. The problem of this approach is that we have the term ω̅/|ω̅| in our constraint. This could result in an division-by-zero problem when ω̅ is small (which is often the case, especially we typically initialize the trajectory with zero velocity). An alternative is to introduce slack variables `θ, a`, where `a` is a 3 x 1 vector with constraint
```
θ*a = ω̅*dt
aᵀa = 1
z[n+1] = [cos(θ/2) sin(θ/2)*a] ⊗ z[n]
```
With these slack variables and additional constraint, we avoid the division-by-zero problem.

cc @RussTedrake

**Comment by hongkai-dai** [comment]
I think I figured out a way to impose the quaternion integration constraint without introducing new variables. Here is the idea

We still want to impose the constraint
```
z[n+1] = [cos(θ/2) sin(θ/2)*a] ⊗ z[n]       (1)
```
where we know θ = |ω̅ * dt|, a = ω̅ / |ω̅|. 

Here we use the condition
```
z[n+1] ⊗ z[n]* = [cos(θ/2) sin(θ/2)*a]              (2)
```
where `z[n]*` is the conjugate of the quaternion `z[n]`. We want to impose the constraint that the left-hand side equals to the right hand-side, without using `θ` or `a`.

First notice that cos(θ/2) = cos(|ω̅ * dt|/2), hence we have the constraint
```
(z[n+1] ⊗ z[n]* )[0] = cos(|ω̅ * dt|/2)            (3)
```
We also have the condition that `(z[n+1] ⊗ z[n]*).tail<3>() ` is in the same direction as `a` (since the bottom 3 entries of right-hand side of equation (2) is sin(θ/2)*a, and sin(θ/2) > 0 as θ > 0), we have the condition
```
cross((z[n+1] ⊗ z[n]*).tail<3>(), ω̅ * dt) = 0                    (4a)
dot((z[n+1] ⊗ z[n]*).tail<3>(), ω̅ * dt) >= 0                     (4b)
```
These two conditions (4a) and (4b) guarantee that (z[n+1] ⊗ z[n]*).tail<3>() is in the same direction as `a`, even when ω̅ * dt = 0.

Hence we only need to enforce the constraint (3), (4a), (4b), without introducing any slack variables.

The caveat is that in condition (3), the term |ω̅ * dt| is NOT differentiable at 0.

**Comment by hongkai-dai** [comment]
It turns out that although the constraint proposed in https://github.com/RobotLocomotion/drake/issues/14956#issuecomment-827857257 doesn't have the division-by-zero problem, the solvers don't like it. I implemented that constraint formulation in the PR #14960, and SNOPT/IPOPT has to try many initial guesses to find a solution satisfying that constraint.

On the other hand, by just replacing the term ω/|ω| with ω/(|ω|+ε), the following constraint
```
(z[n+1]⊗ z[n]*) • Δz=1
where Δz = [cos(|ω|h/2), ω/(|ω|+ε)sin(|ω|h/2)]
```
both SNOPT and IPOPT are much more likely to find a solution.

**Comment by hongkai-dai** [comment]
Resolved already in #14960
